src/cheri_pte.sail

```diff
@@ -38,102 +38,99 @@ union PTE_Check = {
-    (Execute(),       Supervisor) => to_pte_check(p.U() == 0b0 & p.X() == 0b1),
+    (Read(Data),      Supervisor) => ((p.U() == 0b0 | do_sum) & (p.R() == 0b1 | (p.X() == 0b1 & mxr)), ext_ptw),
+    (Write(Data),     Supervisor) => ((p.U() == 0b0 | do_sum) & p.W() == 0b1, ext_ptw),
+    (ReadWrite(Data), Supervisor) => ((p.U() == 0b0 | do_sum) & p.W() == 0b1 & (p.R() == 0b1 | (p.X() == 0b1 & mxr)), ext_ptw),
+    (Execute(),       Supervisor) => (p.U() == 0b0 & p.X() == 0b1, ext_ptw),
 
     (Read(Cap),       User)       => { /* load-cap should probably not be affected by mxr. */
                                        if   p.U() == 0b1 & p.R() == 0b1
                                        then {
                                          let  e = Mk_Ext_PTE_Bits(ext);
                                          if   e.LoadCap() == 0b1
-                                         then (true, PTE_CAP_OK)
+                                         then (true, ext_ptw)
                                          else {
                                            /* Allow the address translation to proceed, but mark for tag stripping */
-                                           (true, PTE_LOAD_CAP_ERR)
+                                           (true, ext_ptw_join_lcm(ext_ptw, PTW_LCM_CLR))
                                          }
                                        }
-                                       else (false, PTE_CAP_OK)
+                                       else (false, ext_ptw)
                                      },
 
     (Write(Cap),      User)       => { if   p.U() == 0b1 & p.W() == 0b1
                                        then {
                                          let  e = Mk_Ext_PTE_Bits(ext);
                                          if   e.StoreCap() == 0b1
-                                         then (true, PTE_CAP_OK)
+                                         then (true, ext_ptw)
                                          else {
                                            /* Do not allow the address translation to proceed */
-                                           (false, PTE_STORE_CAP_ERR)
+                                           (false, ext_ptw_join_scm(ext_ptw, PTW_SCM_TRAP))
                                          }
                                        }
-                                       else (false, PTE_CAP_OK)
+                                       else (false, ext_ptw)
                                      },
 
     (ReadWrite(Cap),  User)       => { if   p.U() == 0b1 & p.R() == 0b1 & p.W() == 0b1
                                        then {
                                          let  e = Mk_Ext_PTE_Bits(ext);
                                          if   e.StoreCap() == 0b1 & e.LoadCap() == 0b1
-                                         then (true, PTE_CAP_OK)
```

Comment:
Should this whole thing not be `ext_ptw_join_scm(ext_ptw_join_lcm(ext_ptw, e.LoadCap() == 0b1 ? PTW_LCM_OK, PTW_LCM_CLR), e.StoreCap() == 0b1 ? PTW_SCM_OK, PTW_SCM_TRAP)`, with suitable local variables of course to make that more readable? This is currently a bug I believe:

```
cmove ca1, cnull
amoswap.c ca1, (a0)
```

Currently, that will allow you to bypass the tag clearing loads, as we'll just get back the `PTW_SCM_TRAP` and then not trap because the value being stored isn't tagged.

For the Read/Write cases I'd do similar things and push the conditional inside the argument of `ext_ptw_join_[ls]cm`, rather than conditionally call it. I think that's neater.

Reply:
Yes, that's probably right.  Standby.

Reply:
Mm, `amoswap.c` doesn't seem to be implemented, and I don't know that it would have hit that case even so, since it should still have issued a `ReadWrite(Cap)` and `PTW_SCM_TRAP` is always paired with a failing translation, but I think it's cleaner your way anyway. :)